akka-stream-tests/src/test/scala/akka/stream/scaladsl/FutureFlattenSpec.scala

```diff
@@ -0,0 +1,219 @@
+
+        val sub = c.expectSubscription()
+        sub.request(5)
+
+        c.expectNext(1)
+        c.expectNext(2)
+        c.expectNext(3)
+
+        c.expectComplete()
+
+        materialized.future.futureValue(Timeout(3.seconds)) should ===("foo")
+      }
+    }
+
+    "use a materializer without auto-fusing" when {
+      implicit val noFusing = ActorMaterializer(
+        ActorMaterializerSettings(system).withAutoFusing(false))
+      implicit def ec = noFusing.executionContext
+
+      val tooDeepForStack = 50000
+
+      "flattening from a future graph" in assertAllStagesStopped {
+        val g = Source.fromFutureSource(Future {
+          Thread.sleep(2000)
+          Fusing.aggressive((1 to tooDeepForStack).
+            foldLeft(Source.single(42).mapMaterializedValue(_ ⇒ 1))(
+              (f, i) ⇒ f.map(identity)))
+        })
+
+        val (mat, fut) = g.toMat(Sink.seq)(Keep.both).run()
+        mat.futureValue(veryPatient) should ===(1)
+        fut.futureValue(veryPatient) should ===(List(42))
+      }
+
+      "flattening from a completion stage" in assertAllStagesStopped {
+        val future: Future[Graph[SourceShape[Int], Int]] = Future {
+          Fusing.aggressive((1 to tooDeepForStack).
+            foldLeft(Source.single(43).mapMaterializedValue(_ ⇒ 1))(
+              (f, i) ⇒ f.map(identity)))
+        }
+        val stage: CompletionStage[Graph[SourceShape[Int], Int]] = future.toJava
+        val g = Source.fromSourceCompletionStage(stage)
+
+        val (mat, fut) = g.toMat(Sink.seq)(Keep.both).run()
+        mat.toScala.futureValue(veryPatient) should ===(1)
+        fut.futureValue(veryPatient) should ===(List(43))
+      }
+    }
+
+    "be cancelled before the underlying Future completes" in {
+      implicit def m = materializer
+
+      assertAllStagesStopped {
+        val promise = Promise[Source[Int, Int]]()
+        val aside = Promise[Int]()
+        val result = Promise[akka.Done]()
+        def futureSource = Source.fromFutureSource(
+          promise.future).map { i ⇒
+          aside.success(i); i // should never occur
+        }.watchTermination[Unit]() {
+          case (_, res) ⇒ result.completeWith(res); ()
+        }
+
+        futureSource.runWith(Sink.cancelled) should ===(NotUsed)
+        result.future.futureValue should ===(akka.Done)
+        aside.future.isCompleted should ===(false)
+      }
+    }
+
+    "fails as the underlying Future is failed" in {
+      implicit def m = materializer
+
+      assertAllStagesStopped {
+        val promise = Promise[Source[Int, Int]]()
+        val result = Promise[akka.Done]()
+        def futureSource = Source.fromFutureSource(promise.future)
+        def sink = Sink.fold[Int, Int](1)(_ * _)
+
+        promise.failure(new Exception("Foo"))
+
+        futureSource.runWith(sink).failed.
+          map(_.getMessage)(m.executionContext).futureValue should ===("Foo")
+      }
+    }
+
+    "applies back-pressure according future completion" in {
```

Comment:
subscription requests 5 elements on line 131 so after that there is demand already, it then completes the future source on line 133, so there is demand when the future completes - no back pressure applied.

For me to make this test case different from what is already covered you would:
1. put a `Future.successful(TestPublisher.probe)` in your stage in a stream with a test subscriber probe downstream (I don't think completing it on a different thread matters)
2. run the stream
3. see that the publisher has gotten no demand (since downstream is backpressuring)
4. demand from downstream and emit from the publisher to show that there was no failure when the source was completed but there was no demand

I'd also rename it to something along the lines of "handle that there is back pressure when the future completes"